raiden/network/transport/matrix/utils.py

```diff
@@ -28,6 +43,178 @@
 ROOM_NAME_PREFIX = 'raiden'
 
 
+class UserPresence(Enum):
+    ONLINE = 'online'
+    UNAVAILABLE = 'unavailable'
+    OFFLINE = 'offline'
+    UNKNOWN = 'unknown'
+
+
+class AddressReachability(Enum):
+    REACHABLE = 1
+    UNREACHABLE = 2
+    UNKNOWN = 3
+
+
+USER_PRESENCE_REACHABLE_STATES = {UserPresence.ONLINE, UserPresence.UNAVAILABLE}
+USER_PRESENCE_TO_ADDRESS_REACHABILITY = {
+    UserPresence.ONLINE: AddressReachability.REACHABLE,
+    UserPresence.UNAVAILABLE: AddressReachability.REACHABLE,
+    UserPresence.OFFLINE: AddressReachability.UNREACHABLE,
+    UserPresence.UNKNOWN: AddressReachability.UNKNOWN,
+}
+
+
+class UserAddressManager:
```

Comment:
So in the services we'd instantiate a `UserAddressManager` based on the `GMatrixClient` and provide proper callbacks to update the routing table based on that.

What's the difference between address reachability and user presence though and/or which one should we prefer.


Besides that the PR looks good.

Reply:
> So in the services we'd instantiate a UserAddressManager based on the GMatrixClient and provide proper callbacks to update the routing table based on that.

Exactly.
One important caveat: In order for presence tracking to occur addresses need to be 'registered' with the `UserAddressManager` by calling the `add_address()` method. 

Question: Would it be useful for you to have an `add_addresses()` method that takes an iterable so you can add all addresses you care about after a restart in one go?

> What's the difference between address reachability and user presence though and/or which one should we prefer.

For our purposes an `address` here is basically equivalent to a Raiden node.
One node with one `address` can be represented by multiple `user`s on multiple homeservers.
In the Raiden transport we need to know about individual user presence in order to react to changes in that (e.g. invite 'new' users for already known addresses into existing communication rooms).

In the services you shouldn't need to care about users, only addresses. That means you also don't need to provide a user presence callback.

I'll add a docstring to the class explaining this.

Reply:
> Question: Would it be useful for you to have an add_addresses() method that takes an iterable so you can add all addresses you care about after a restart in one go?

The participants are stored on a per channel basis, so this wouldn't make a big improvement.

> I'll add a docstring to the class explaining this.

Thanks for the explanation.